CIRC-7499 - IRONdb 0.21.0 Docs (#320)
circonus/docs: content/irondb/release-notes.md

# Release Notes

## Changes in 0.21.0

2022-01-07

>  ** IMPORTANT: This update changes the format of the metric name database.
>      The database will automatically be converted to the new format the
>      first time the software boots. Once this has been done, the software
>      cannot be reverted to previous versions unless you wipe out the contents
>      of the metric name database first. If you need to downgrade for any reason
>      after updating to this version, please contact Circonus support **

Reviewer: ```suggestion

  **IMPORTANT: This update changes the format of the metric name database.
   The database will automatically be converted to the new format the
   first time the software boots. Once this has been done, the software
   cannot be reverted to previous versions unless you wipe out the contents
   of the metric name database first. If you need to downgrade for any reason
   after updating to this version, please contact Circonus support.**
```

Reviewer: This fixes line break after the date and the bolding, which requires the `**` to not be separated by whitespace at the start and end.

>  * Remove deprecated `nnt` field from `/state` output

Reviewer: ```suggestion
 * Remove deprecated `nnt` field from `/state` output.
```

 * Fix potential crash when bad data points are found during raw data rollups
 * Fix `tag_cats` and `tag_vals` endpoints to respect the `X-Snowth-Advisory-Limit` header.
 * Added ability to use relative time-based values for activity fields in `/find` queries.
   Previously, the `activity_start_secs` and `activity_end_secs` fields had to be provided

>    in epoch seconds. The string `now` will be accepted as a field. It will user the current

Reviewer: ```suggestion
   in epoch seconds. The string `now` will be accepted as a field. It will use the current
```

   time. You may also use values like `+10s` to indicate "ten seconds from now' or `-10s` to
   indicate "ten seconds ago". For example: a query that reads
   `activity_start_secs=-1d&activity_end_secs=now` will pull records that have activity data
   in the last day. A full list of accepted formats for these new values can be found
   [here](https://circonus-labs.github.io/libmtev/config/time_durations.html)

>  * Improve metric database loading times

Reviewer: ```suggestion
 * Improve metric database loading times.
```

>  * Improve CAQL error messages

Reviewer: ```suggestion
 * Improve CAQL error messages.
```

 * Fix bug where find requests were not respecting the default limit if no limit

>    was manually specified

Reviewer: ```suggestion
   was manually specified.
```

 * `graphite_translate` additions/fixes to better match graphite behavior:
   * new functions - `add()`, `movingMin()`, `randomWalk()`, `stddevSeries()`, `pow()` and `powSeries()`

>    * `pieMin()`, pieAverage() and pieMax() implemented as min(), avg() and max()

Reviewer 2: ```suggestion
   * `pieMin()`, `pieAverage()` and `pieMax()` implemented as `min()`, `avg()` and `max()`
```

   * all sorting and pictire manipulation functions implemented as pass()

>  * Fix race condition on old shard removal that can cause corruption

Reviewer: ```suggestion
 * Fix race condition on old shard removal that can cause corruption.
```

>  * Improve `find` query performance in some pathological cases

Reviewer: ```suggestion
 * Improve `find` query performance in some pathological cases.
```

 * Fix possible crash when deleting a timeshard while it's also being

>    resized

Reviewer: ```suggestion
   resized.
```

>  * Fix graphite `irondb_tag_filter` search bug that prevented query proxying

Reviewer: ```suggestion
 * Fix graphite `irondb_tag_filter` search bug that prevented query proxying.
```

>  * Fix potential crash on when invalid histogram journal data is received by a node

Reviewer: ```suggestion
 * Fix potential crash on when invalid histogram journal data is received by a node.
```

Author: Fixed manually - also fixed a typo ("on when" -> "when")

>  * Fix races in arts and arts compaction

Reviewer: ```suggestion
 * Fix races in arts and arts compaction.
```

Author: I just removed this line altogether. It's not useful for a customer.

>  * Reconstitute recovery from errors on the sending nodes has been partially automated,
>    now automatically skips shards on errors with the sending node after a configurable
>    number of retries and will once more retry them after finishing all the well-behaving
>    shards, then providing a final list of shards in the errorlog and on disk that could
>    not be successfully reconstituted after the reconstitute is finished (allowing further
>    diagnosis and actions to be taken by the operator as needed). Shards which are in
>    maintenance will also be skipped and retried in a similar way

Reviewer: ```suggestion
 * Reconstitute recovery from errors on the sending nodes has been partially
   automated. It now automatically skips shards on errors from the sending node
   after a configurable number of retries and will retry them once more after
   finishing all the well-behaving shards. Upon completion, a final list of
   shards that could not be successfully reconstituted will be recorded in the
   `errorlog`, allowing further diagnosis and actions to be taken by the
   operator as needed. Shards which are in maintenance will also be skipped and
   retried in a similar way.
```

This is my rewrite for the Inside release notes. I found the original to be difficult to follow.

 * Fix `/find/X/active_count` to respect `query` querystring parameter and to force
   `activity_start_secs` and `activity_end_secs` into activity window alignment.

>  * Allow `graphite/find` to accept an optional `irondb_tag_filter` query string parameter

Reviewer: ```suggestion
 * Allow `graphite/find` to accept an optional `irondb_tag_filter` query string parameter.
```

   This parameter will take a circonus-style tag-query and filter the metrics down to only include matching metrics.

>  * Fix bug in rollup scheduler that could cause rollups to get scheduled to run multiple times

Reviewer: ```suggestion
 * Fix bug in rollup scheduler that could cause rollups to get scheduled to run multiple times.
```

>  * Improvements to reconstitute stability

Reviewer: ```suggestion
 * Improve reconstitute stability.
```

>  * Fix incorrectly formatted error message when receiving invalid graphite input

Reviewer: ```suggestion
 * Fix incorrectly formatted error message when receiving invalid graphite input.
```

>  * Fix potential crash on bad fetches from CAQL

Reviewer: ```suggestion
 * Fix potential crash on bad fetches from CAQL.
```

 * Fix bug in reconstitute that caused unnecessary replaying of data, slowing down the recovery
   process after a node finishes reconstituting.

>  * Improved `/fetch` error handling in CAQL

Reviewer: ```suggestion
 * Improved `/fetch` error handling in CAQL.
```
